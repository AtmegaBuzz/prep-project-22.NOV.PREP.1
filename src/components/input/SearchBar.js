import { useEffect, useRef, useState } from "react";
import { getAutocompleteCities } from "../../utils";

const SearchBar = ({ city, setCity }) => {
  const inputRef = useRef();
  const [autocompleteCities, setAutocompleteCities] = useState([]);
  let reqTimeout = null;

  useEffect (() => {
    inputRef.current.value = city;
  }, [city]);

  // Load suggestations
  const handleOnChange = async (e) => {
    if (reqTimeout) {
      clearTimeout(reqTimeout);
    }

    const cityValue = inputRef.current.value;

    reqTimeout = getAutocompleteCities(cityValue, setAutocompleteCities);
  };

  // If user clicks enter on input field
  const handleEnter = async (e) => {
    const cityValue = inputRef.current.value;

    // Triggered when user clicks enter on input field
    if (e.code === "Enter") {
      setCity(cityValue);
    }
  };

  // If user clicks on item from data list
  const handleListItemClick = async (e) => {
    const value = e.currentTarget.value;

    autocompleteCities.forEach((city) => {
      if (city.city === value) {
        setCity(city.city);
        return;
      }
    });
  };

  return (
    <div className="input-container">
      <i className="fa fa-map-marker input-icon" aria-hidden="true"></i>
      <input
        data-testid="search-bar"
        ref={inputRef}
        list="cities"
        name="search"
        type="text"
        defaultValue="New York City"
        placeholder="Enter a city"
        autoComplete="off"
        onKeyDown={handleEnter}
        onChange={handleOnChange}
        onInput={handleListItemClick}
      />

      {
        autocompleteCities.length > 0 && (
          <datalist data-testid="data-list" id="cities">
            {autocompleteCities.map((city, i) => (
              <option
                key={i}
                value={city.city}
              >{`${city.city}, ${city.country}`}</option>
            ))}
          </datalist>
        )
      }
    </div>
  );
};

export default SearchBar;
