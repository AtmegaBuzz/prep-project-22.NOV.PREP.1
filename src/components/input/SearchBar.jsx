const SearchBar = ({ inputRef, onKeyDown, onChange, autocompleteCities }) => {
  return (
    <div className="input-container">
      <i className="fa fa-map-marker input-icon" aria-hidden="true"></i>
      <input
        list="cities"
        name="search"
        type="text"
        ref={inputRef}
        onKeyDown={onKeyDown}
        onChange={onChange}
        defaultValue="New York City"
        placeholder="Enter a city"
        autoComplete="off"
      />

      <datalist id="cities">
        {autocompleteCities.map((city, i) => (
          <option key={i}>{`${city}`}</option>
        ))}
      </datalist>
    </div>
  );
};

export default SearchBar;
