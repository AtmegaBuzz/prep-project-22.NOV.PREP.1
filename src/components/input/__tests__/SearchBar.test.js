import { render, screen, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import SearchBar from "../SearchBar";

afterEach(() => {
    cleanup();
    }
);

test("renders SearchBar component", () => {
    render(<SearchBar />);
    const searchBarElement = screen.getByTestId("search-bar");
    expect(searchBarElement).toHaveAttribute("placeholder");
    expect(searchBarElement).toHaveAttribute("type", "text");
    expect(searchBarElement).toBeInTheDocument();
})

test("match snapshot", () => {
    const tree = renderer.create(<SearchBar />).toJSON();
    expect(tree).toMatchSnapshot();
})