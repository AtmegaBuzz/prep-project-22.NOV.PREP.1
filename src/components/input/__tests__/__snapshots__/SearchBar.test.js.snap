// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match snapshot 1`] = `
<div
  className="input-container"
>
  <i
    aria-hidden="true"
    className="fa fa-map-marker input-icon"
  />
  <input
    autoComplete="off"
    data-testid="search-bar"
    defaultValue="New York City"
    list="cities"
    name="search"
    onChange={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    placeholder="Enter a city"
    type="text"
  />
</div>
`;
